proposals/0724-qualified-lists.rst

```diff
@@ -0,0 +1,395 @@
+      - ``M.buildListEnum (Data.List.Qualified.Experimental.EnumFromThenTo x y z)``
+
+.. list-table::
+    :align: left
+
+    * - **New pattern syntax**
+      - **Desugared pattern syntax**
+    * - ``M.[x, _, y]``
+      - ``M.FromListCons x (M.FromListCons _ (M.FromListCons y M.FromListNil))``
+
+``Data.List.Qualified.Experimental`` will initially live in ``ghc-experimental``, eventually merged into ``GHC.Exts``. It will contain:
+
+::
+
+  data EnumFrom from then to
+    = EnumFrom from
+    | EnumFromThen from then
+    | EnumFromTo from to
+    | EnumFromThenTo from then to
+
+It is highly recommended that all types with ``IsList`` instances defined provide a ``.Interpolate`` module with the below definitions, to enable locally-scoped overloading over ``-XOverloadedLists``, for example:
+
+::
+
+  import GHC.Exts qualified as L
+  import Prelude qualified
+
+  buildList :: ((a -> a -> [a]) -> [a] -> [a]) -> MyList a
+  buildList f = L.fromList (f (:) [])
+
+  buildListEnum :: EnumFrom a a a -> MyList a
+  buildListEnum e =
+    case e of
+      EnumFrom x -> L.fromList (Prelude.enumFrom x)
+      EnumFromThen x y -> L.fromList (Prelude.enumFromThen x y)
+      EnumFromTo x y -> L.fromList (Prelude.enumFromTo x y)
+      EnumFromThenTo x y z -> L.fromList (Prelude.enumFromThenTo x y z)
+
+Note that while we could have mirrored ``-XOverloadedLists`` and just done ``M.fromListN 2 [x, y]``, we intentionally decide to use this more general API. This gives us more expressive power, since we no longer need to typecheck an intermediate list. Similar reason for defining ``buildListEnum`` instead of reusing Prelude's ``enumFrom`` functions. See *Section 4.2 Heterogeneous Lists* for a use-case.
+
+We also decide to do ``M.buildList`` instead of something like ``M.fromList (x `M.cons` M.nil)`` so that there's one definition to jump to (e.g. with IDE integrations) instead of three.
+
+To use as patterns, the implementor should define ``FromListCons`` and ``FromListNil`` pattern synonyms, typically with the ``COMPLETE`` pragma specified. We choose to do this instead of ``toList -> [x, _, z]`` because that would also disallow heterogeneous lists.
+
+Parser
+~~~~~~
+
+Update `Section 10.5 <https://www.haskell.org/onlinereport/haskell2010/haskellch10.html#x17-18000010.5>`_ of the Haskell 2010 report as follows.
+
+.. code-block:: abnf
+
+  aexp → qvar
+       | ...
+       | modid . [ exp_1 , ..., exp_k ]
+
+  apat → var [ @ apat ]
+       | ...
+       | modid . [ pat_1 , ..., pat_k ]
+
+Module name resolution
+~~~~~~~~~~~~~~~~~~~~~~
+
+Module names are resolved immediately, when parsing a quote. This matches the behavior of resolving modules in normal qualified values in quotes.
+
+::
+
+  module A where
+
+  import OneImpl qualified as M
+
+  -- Immediately resolves to OneImpl.[1, 2]
+  -- Errors if M is not in scope
+  foo = [| M.[1, 2] |]
+
+::
+
+  module B where
+
+  import AnotherImpl qualified as M
+  import A
+
+  bar = $foo
```

Comment:
it's not clear to me what point you're trying to make with module B

Reply:
To show usage of module A, but it's unnecessary, I'll remove